app/service/operation_svc.py

```diff
@@ -91,9 +93,32 @@ def __init__(self):
             if agent_output:
                 result = (await self.data_svc.explode_results(criteria=dict(link_id=step['id'])))[0]
                 step_report['output'] = self.decode_bytes(result['output'])
-            report['steps'].append(step_report)
+            agents_steps[step['paw']]['steps'].append(step_report)
+        report['steps'] = agents_steps
+        report['skipped_abilities'] = await self.get_skipped_abilities_by_agent(op_id=op['id'])
         return report
 
+    async def get_skipped_abilities_by_agent(self, op_id):
+        """
+        Generate a list of skipped abilities for agents in an operation
+        :param op_id: operation id
+        :return: a JSON skipped abilities list by agent
```

Comment:
I don't like the explosion of this service class when we beef up the reporting like this. can you extract this operation report piece into a separate module?

Reply:
Do you mean creating a new `reporting_svc`?